datastore/storagebackend/postgresql/gettsattrgroups.go

```diff
@@ -0,0 +1,326 @@
+// otherwise (false, nil). If an error occurs, (..., error) is returned.
+func tsMdataEqual(tsmd1, tsmd2 *datastore.TSMetadata, goNames []string) (bool, error) {
+	tp1 := reflect.ValueOf(tsmd1)
+	tp2 := reflect.ValueOf(tsmd2)
+
+	for _, goName := range goNames {
+		field1 := tp1.Elem().FieldByName(goName)
+		if !field1.IsValid() {
+			return false, fmt.Errorf("tp1.Elem().FieldByName(%s) returned invalid value", goName)
+		}
+		field2 := tp2.Elem().FieldByName(goName)
+		if !field2.IsValid() {
+			return false, fmt.Errorf("tp2.Elem().FieldByName(%s) returned invalid value", goName)
+		}
+
+		if !field1.Equal(field2) {
+			return false, nil // at least one difference found
+		}
+	}
+
+	return true, nil // no differences found
+}
+
+// getCombo gets a version of tsMdata1 where only the given attributes (in goNames) have been set
+// (all other attributes having the default values).
+// Returns (TSMetadata, nil) upon success, otherwise (..., error).
+func getCombo(tsMdata1 *datastore.TSMetadata, goNames []string) (*datastore.TSMetadata, error) {
+	tsMdata2 := datastore.TSMetadata{}
+	tp1 := reflect.ValueOf(tsMdata1)
+	tp2 := reflect.ValueOf(&tsMdata2)
+
+	for _, goName := range goNames {
+		field1 := tp1.Elem().FieldByName(goName)
+		if !field1.IsValid() {
+			return nil, fmt.Errorf("invalid field (1) (goName: %s)", goName)
+		}
+		field2 := tp2.Elem().FieldByName(goName)
+		if !field2.IsValid() {
+			return nil, fmt.Errorf("invalid field (2) (goName: %s)", goName)
+		}
+		if !field2.CanSet() {
+			return nil, fmt.Errorf("unassignable field (2) (goName: %s)", goName)
+		}
+
+		switch field1.Kind() {
+		case reflect.String:
+			field2.SetString(field1.String())
+		default:
+			return nil, fmt.Errorf("unsupported type: %v (goName: %s)", field1.Kind(), goName)
+		}
+	}
+
+	return &tsMdata2, nil
+}
+
+// getTSAttrGroupsIncInstances populates groups from cols such that each group contains all
+// instances that match a unique combination of database values corresponding to cols.
+// All attributes, including those in cols, are set to the actual values found in the database.
+// Returns nil upon success, otherwise error.
+func getTSAttrGroupsIncInstances(
+	db *sql.DB, cols []string, groups *[]*datastore.TSMdataGroup) error {
+	allCols := getTSAllPBNames() // get all protobuf names of TSMetadata message
+
+	goNames, err := getTSGoNamesFromPBNames(cols)
+	if err != nil {
+		return fmt.Errorf("getTSGoNamesFromPBNames() failed: %v", err)
+	}
+
+	// query database for all columns in time_series, ordered by cols
+	allColsS := strings.Join(allCols, ",")
+	colsS := strings.Join(cols, ",")
+	query := fmt.Sprintf("SELECT %s FROM time_series ORDER BY %s", allColsS, colsS)
+	rows, err := db.Query(query)
+	if err != nil {
+		return fmt.Errorf("db.Query() failed: %v", err)
+	}
+	defer rows.Close()
+
+	// aggregate rows into groups
+	currInstances := []*datastore.TSMetadata{} // initial current instance set
+	for rows.Next() {
```

Comment:
It is a pity you can't easily use `TSMetadata` as a map key, as I think putting them in a map first would clean up this code.
But with that limitation, what you did is a neat solution.

Reply:
I thought TSMetadata can't be used as a key because it doesn't implement `==` and `!=`. I tried once with making the key `interface{}`, but that seemed to strange. But what do I know...

Reply:
What I had in mind is to fill a map where `combos` is the key, and the value is the list of `instances`. Then you can fill the groups at the end from that. I think it removes some of the duplication in the code (but your approach is more memory efficient).
Hard to see if it would actually look easier without actually doing it...

Reply:
I was too quick above, since I was thinking of using `*datastore.TSMdataGroup` as the map key, which works since a pointer is always a comparable type (i.e. implementing == and !=)! But we can't use pointers in this case, and would therefore need to somehow make `datastore.TSMdataGroup` itself comparable. Not sure how to achieve that - especially since TSMdataGroup is a generated type. Maybe someone has written a special protoc that makes types comparable, but I think we just leave that exercise for later. Feel free to investigate :)